### ISSUE TITLE
Formatting Quantity and other mixin columns

### ISSUE THREAD

**Issue by taldcroft** [astropy/astropy#5265]
The current mixin formalism provides only a way to specify a default column format via the `info` attribute.  For instance:

```
class QuantityInfo(ParentDtypeInfo):
    """
    Container for meta information like name, description, format.  This is
    required when the object is used as a mixin column within a table, but can
    be used as a general way to store meta information.
    """
    attrs_from_parent = set(['dtype', 'unit'])  # dtype and unit taken from parent
    _supports_indexing = True

    @staticmethod
    def default_format(val):
        return '{0.value:}'.format(val)
```

But as noted in #5263 this does not provide the ability to set the `info.format` attribute in a way that should work, e.g. `dat['a'].info.format = '%f'`.  This ends up failing here (in `table/pprint.py`) because `'%f' % quantity` fails:

```
def _possible_string_format_functions(format_):
    """Iterate through possible string-derived format functions.

    A string can either be a format specifier for the format built-in,
    a new-style format string, or an old-style format string.
    """
    yield lambda format_, val: format(val, format_)
    yield lambda format_, val: format_.format(val)
    yield lambda format_, val: format_ % val
```

In the particular case of `Quantity`, everything would work just dandy if `val` were actually `val.value`.  That suggests one potential generalization of allowing for a method in `info` that transforms from the actual object being formatted to `val`.  Another idea to handle this is providing overrides for each of the above three string format functions.  This is probably more general and powerful, but comes with a performance penalty that might be non-trivial.  Maybe we could make two versions of `_possible_string_format_functions`, one which gets called if there are info overrides and the other (equivalent to the current version) if there are no info overrides.

Thoughts @mhvk?

**Comment by mhvk** [comment]
@taldcroft - it needs a bit of thought, though perhaps a start would be to use new-style formats everywhere, since for those `Quantity.__format__` will do some of the work for us:
```
q = 2.*u.m'
%12.4f' % q
# TypeError: Only dimensionless scalar quantities can be converted to Python scalars
'{:12.4f}'.format(q)
# '      2.0000 m'
```
I think it may make sense in `ecsv` files to give just the format without any percent of curly braces; everything else is set up to deal with that!

Now the other problem with the example in #5263 is that the type is defined as integer, while `Quantity` automatically converts to `float`, and thus cannot be properly represented with `d`. Indeed, if I do:
```
ecsv_content = """# %ECSV 0.9
# ---
# datatype:
# - {name: a, unit: m / s, datatype: int64, format: '%03d'}
# - {name: b, unit: km, datatype: int64, description: This is column b}
a b
001 2
004 3
"""
from astropy.table import QTable
qt = QTable.read(ecsv_content, format='ascii.ecsv')
qt['a'].info.format = '{:12.4f}'
qt
# <QTable length=2>
#         a             b   
#       m / s           km  
#      float64       float64
# ------------------ -------
#       1.0000 m / s     2.0
#       4.0000 m / s     3.0
```
which at least works, though it is not that great either in that we do not want the unit on every line.

Fortunately, we did think things through reasonably well in the pretty printer, and were flexible enough to allow `format` to be a callable, so the following does work!
```
qt['a'].info.format = '{0.value:12.4f}'.format
qt
# <QTable length=2>
#      a          b   
#    m / s        km  
#   float64    float64
#  ------------ -------
#       1.0000     2.0
#       4.0000     3.0
```

This suggests that a solution to the above conundrum is to make `QuantityInfo` have `format` as a property that has a setter which changes any simple format string (like `'12.4f'`) to `'{0.value:<str>}'.format`. What do you think? This would seems reasonably straightforward to implement...

p.s. Two separate and possibly more difficult issues are:
1. When data is read with a specific type specified, `QTable` and thus `Quantity` should respect that; this may be as easy as passing on the `datatype` as a `dtype` in the initialiser, but I haven't quite looked at how it is done.
2. It is a bit unfortunate that if a bad format is present, the table cannot be printed anymore... Rather than raising an exception, should we print a warning and just fill the guilty column with *** or so??

**Comment by mhvk** [comment]
p.s. Even just setting `qt['a'].info.format = '{0.value:12.4f}'` works, so all that would be required is some string-wrangling in the `format` setter.

**Comment by mhvk** [comment]
fixed by #5274, so closing
